Source Notion: title text can be an array of object

mdibaiee opened the pull request:
## What
- Source Notion's schema currently specifies title text as an array, but it can also be an object. Here is an example document I observed:
```
{
  "id": "b649a924-70fa-43c3-812e-262b4043ae27",
  "properties": [
    {
      "name": "Name",
      "value": {
        "id": "title",
        "title": [
          {
            "annotations": {
              "bold": false,
              "code": false,
              "color": "default",
              "italic": false,
              "strikethrough": false,
              "underline": false
            },
            "href": null,
            "plain_text": "Europe Trip",
            "text": {
              "content": "Europe Trip",
              "link": null
            },
            "type": "text"
          }
        ],
        "type": "title"
      }
    }
  ]
}
```

## How
- Accept both an array of objects, or a single object as `title.text`

## Recommended reading order
1. `airbyte-integrations/connectors/source-notion/source_notion/schemas/shared/title.json`
2. `airbyte-integrations/connectors/source-notion/source_notion/schemas/shared/title_text.json`

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

Not a breaking change.

## Pre-merge Checklist

<strong>Updating a connector</strong>

### Community member or Airbyter

- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [x] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

## Tests

<details><summary><strong>Unit</strong></summary>

<img width="707" alt="image" src="https://user-images.githubusercontent.com/2807772/202200963-322d0541-82ff-44f5-ae29-66805f38c876.png">

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

marcosmarxm commented:
Hello :wave:, first thank you for this amazing contribution. 

We really appreciate the effort you've made to improve the project.
We ask you patience for the code review. Last month our team was focused on Hacktoberfest event and that probably left some PR without the proper feedback. And this week, due to the Thanksgiving US Holiday, most our team is out of office with their families. Another important piece of information why code won't be merge this week is: as a safety measure the core team has decided to freeze merging code to main branch to keep the release stable. Next week we'll return to you with the proper code review and update the status of your contribution.

If you have any questions feel free to send me a message in Slack!
Thanks!

marcosmarxm commented:
/test connector=connectors/source-notion

marcosmarxm commented:
/test connector=connectors/source-notion

marcosmarxm commented:
Hello :wave::skin-tone-2: and thank you for your contribution!

Airbyte has instituted a code freeze between 19 and 30 December, to make sure there are no disruptions during the holidays. 
Because of this, reviewing and merging your contribution may take longer than usual. 
We apologize for the delay, but we want everyone to have a quiet and happy holiday.

If you have any questions or need further clarification, please don't hesitate to ping via Slack.

mdibaiee commented:
/test connector=connectors/source-notion

mdibaiee commented:
It seems I can't run the tests myself (the bot 👀 my comment but didn't run it). I rebased my changes on master, and fixed an issue with the schema changes and ran the tests locally and they passed for me (although I think some of the integration tests can't be meaningfully run since they require credentials, etc.). Can someone please trigger a run of tests again? Thank you

marcosmarxm commented:
/test connector=connectors/source-notion

mdibaiee commented on airbyte-integrations/connectors/source-notion/source_notion/schemas/shared/title_text.json:
```diff
@@ -0,0 +1,59 @@
+{
+  "type": ["null", "object"],
+  "additionalProperties": true,
+  "properties": {
```
I'm referencing the same schema in two places in `title.json`, to avoid repetition and potential divergence of those two places, I'm using a separate file.

marcosmarxm commented on docs/integrations/sources/notion.md:
```diff
@@ -83,16 +83,17 @@ The connector is restricted by Notion [request limits](https://developers.notion
-| 0.1.1   | 2021-12-30 | [9207](https://github.com/airbytehq/airbyte/pull/9207)   | Update connector fields title/description                       |
-| 0.1.0   | 2021-10-17 | [7092](https://github.com/airbytehq/airbyte/pull/7092)   | Initial Release                                                 |
+|--------|------------|----------------------------------------------------------|-----------------------------------------------------------------|
+| 1.0.1  | 2023-01-12 | [19478](https://github.com/airbytehq/airbyte/pull/19478) | Fix title.text schema to allow single text objects              |
+| 1.0.0  | 2022-12-19 | [20639](https://github.com/airbytehq/airbyte/pull/20639) | Fix `Pages` stream schema                                       |
```
Sorry I missed the previous change it is 1.0.0

marcosmarxm reviewed (approved):
Thanks, I only need approval from the connector team to release this change.

marcosmarxm commented:
@lazebnyi @davydov-d can you do the final review for this change in a GA connector?

davydov-d commented:
@marcosmarxm I'm sorry, I got too much oncall issues to work on these days

davydov-d commented:
hey @mdibaiee I can't spot the difference between your document sample and the schema the connector is exposing now. Can you please double check your data?

mdibaiee commented:
@davydov-d in the current schema `text` is an object or null, whereas the document I have provided an example of, `text` is an array of objects

davydov-d commented:
@mdibaiee you might be missing something, it is an object in your data
![image](https://user-images.githubusercontent.com/7703835/226387805-3fa6cc79-c15d-4842-a699-cb348d892783.png)

mdibaiee commented:
You are right. I tried to look for the original error I was seeing, but this is from 5 months ago and I couldn't find it anymore. 🤷🏽

marcosmarxm commented:
@mdibaiee let us know if you hit the problem again! Sorry the long delay to review the contribution
